src/extensions/surveys.tsx

```diff
@@ -219,3 +118,752 @@ export function generateSurveys(posthog: PostHog) {
+    )
+}
+
+export function Questions({
+    survey,
+    posthog,
+    styleOverrides,
+}: {
+    survey: Survey
+    posthog: PostHog
+    styleOverrides?: React.CSSProperties
+}) {
+    const backgroundColorRef = Preact.createRef()
+    const [textColor, setTextColor] = useState('white')
+    const [questionsResponses, setQuestionsResponses] = useState({})
+    const [currentQuestion, setCurrentQuestion] = useState(0)
+    useEffect(() => {
+        setTextColor(getTextColor(backgroundColorRef.current))
+    }, [])
+
+    const onNextClick = (res: string | string[] | number | null, idx: number) => {
+        const responseKey = idx === 0 ? `$survey_response` : `$survey_response_${idx}`
+        if (idx === survey.questions.length - 1) {
+            sendSurveyEvent({ ...questionsResponses, [responseKey]: res }, survey, posthog)
+            return
+        } else {
+            setQuestionsResponses({ ...questionsResponses, [responseKey]: res })
+            setCurrentQuestion(idx + 1)
+        }
+    }
+    const isMultipleQuestion = survey.questions.length > 1
+
+    return (
+        <form
+            className="survey-form"
+            style={{ color: textColor, borderColor: survey.appearance?.borderColor, ...styleOverrides }}
+            ref={backgroundColorRef}
+        >
+            {survey.questions.map((question, idx) => {
+                if (isMultipleQuestion) {
+                    return (
+                        <>
+                            {currentQuestion === idx && (
+                                <div className={`tab question-${idx} ${question.type}`}>
+                                    {questionTypeMap(
+                                        question,
+                                        idx,
+                                        survey.appearance || defaultSurveyAppearance,
+                                        (res) => onNextClick(res, idx),
+                                        () => closeSurveyPopup(posthog, survey)
+                                    )}
+                                </div>
+                            )}
+                        </>
+                    )
+                }
+                return questionTypeMap(
+                    survey.questions[idx],
+                    idx,
+                    survey.appearance || defaultSurveyAppearance,
+                    (res) => onNextClick(res, idx),
+                    () => closeSurveyPopup(posthog, survey)
+                )
+            })}
+        </form>
+    )
+}
+
+const closeSurveyPopup = (posthog: PostHog, survey: Survey) => {
+    posthog.capture('survey dismissed', {
+        $survey_name: survey.name,
+        $survey_id: survey.id,
+        sessionRecordingUrl: posthog.get_session_replay_url?.(),
+        $set: {
+            [`$survey_dismissed/${survey.id}`]: true,
+        },
+    })
+    localStorage.setItem(`seenSurvey_${survey.id}`, 'true')
+    window.dispatchEvent(new Event('PHSurveyClosed'))
+}
```

Comment:
It would be great if at some point we could move this into propert state management. We don't have to go full Kea.js but some nicer way of encapsulating this logic would be good (also so that we could add unit tests)

Reply:
added as a TODO